katdal/lazy_indexer.py

```diff
@@ -76,6 +76,39 @@ def _simplify_index(shape, indices):
     return tuple(out)
 
 
+def _dask_getitem(x, keep):
+    """Index a dask array, with N-D fancy index support and better performance.
+
+    This is a drop-in replacement for ``x[keep]`` that goes one further
+    by implementing "N-D fancy indexing" which is still unsupported in dask.
+    If `keep` contains multiple fancy indices, perform outer (`oindex`)
+    indexing. This behaviour deviates from NumPy, which performs the more
+    general (but also more obtuse) vectorized (`vindex`) indexing in this case.
+    See  NumPy `NEP 21`_, dask/dask#433 and h5py/h5py#652 for more details.
+
+    .. _NEP 21: http://www.numpy.org/neps/nep-0021-advanced-indexing.html
+
+    In addition, this optimises performance by culling unnecessary nodes from
+    the dask graph after indexing, which makes it cheaper to compute if only a
+    small piece of the graph is needed, and by collapsing fancy indices in
+    `keep` to slices where possible (which also implies oindex semantics).
+    """
+    keep = _simplify_index(x.shape, keep)
+    try:
+        kept = x[keep]
+    except NotImplementedError:
+        # Dask does not like multiple boolean indices: go one dim at a time
+        kept = x
+        for dim, keep_per_dim in enumerate(keep):
+            kept = da.take(kept, keep_per_dim, axis=dim)
```

Comment:
This was suitable for stage1 where the indices were all boolean arrays, but it's inadequate for the general case. Scalar indices, `np.newaxis` and `...` will all need to be handled.